syntax = "proto3";
package geography.proto;

import "market/proto/goods.proto";
import "market/proto/market.proto";
import "industry/proto/industry.proto";

// A unit area, able to hold one production chain.
message Field {
  // Progress of the production chain currently underway.
  industry.proto.Progress progress = 1;
  // Fixed equipment for this field.
  market.proto.Container fixed_capital = 2;
  // Terrain type.
  industry.proto.LandType land_type = 3;
  // Raw materials, for example ores.
  market.proto.Container resources = 4;
  // ID of the field's owner.
  uint64 owner_id = 5;
}

message ResourceLimits {
  // The maximum amount of each raw material in fields in this area.
  market.proto.Container maximum = 1;
  // Recovery speed while a production chain is in progress.
  market.proto.Container recovery = 2;
  // Same while there is no production.
  market.proto.Container fallow_recovery = 3;
}

// A collection of fields close enough together to be treated as a unit, for
// example, to have a single market.
message Area {
  repeated Field fields = 1;
  ResourceLimits limits = 2;
  repeated uint64 pop_ids = 3;
  market.proto.MarketProto market = 4;
  uint64 id = 5;
}

// A series of steps by which a Field can be converted from one LandType to
// another.
message Transition {
  industry.proto.LandType source = 1;
  industry.proto.LandType target = 2;

  // The fixed capital the target will end up with, for example houses.
  market.proto.Container final_fixed_capital = 3;

  // Useful byproducts, e.g. timber from converting a forest.
  market.proto.Container resources_harvested = 4;

  // The amount of fixed capital that requires one additional removal step.
  market.proto.Container step_fixed_capital = 5;

  // Resources required for one step.
  market.proto.Container step_input = 6;

  // Number of steps in the absence of fixed capital.
  int32 steps = 7;
}

// Types of connection.
enum ConnectionType {
  CT_OPEN = 0;
  CT_FOREST = 1;
  CT_MARSH = 2;
  CT_HILLS = 3;
  CT_MOUNTAIN = 4;
  CT_DESERT = 5;
  CT_RIVERINE = 6;
  CT_COASTAL = 7;
  CT_OPEN_SEA = 8;
}

// A connection between two areas.
message Connection {
  ConnectionType type = 1;
  int64 distance = 2;
  int64 width = 3;
  int32 ruggedness = 4;
  int32 chokepoints = 5;

  int64 a_end = 6;
  int64 z_end = 7;
  // TODO: Infrastructure, vulnerability.
}
